x/staking/keeper/msg_server.go

```diff
@@ -649,6 +649,28 @@ func (k msgServer) RotateConsPubKey(ctx context.Context, msg *types.MsgRotateCon
 		return nil, errorsmod.Wrapf(sdkerrors.ErrInvalidType, "expecting cryptotypes.PubKey, got %T", cv)
 	}
 
+	// check if the new public key type is valid
+	paramsRes := consensusv1.QueryParamsResponse{}
+	if err := k.QueryRouterService.InvokeTyped(ctx, &consensusv1.QueryParamsRequest{}, &paramsRes); err != nil {
+		return nil, errorsmod.Wrapf(sdkerrors.ErrInvalidRequest, "failed to query consensus params: %s", err)
+	}
+	if paramsRes.Params.Validator != nil {
+		pkType := pk.Type()
+		if !slices.Contains(paramsRes.Params.Validator.PubKeyTypes, pkType) {
+			return nil, errorsmod.Wrapf(
+				types.ErrValidatorPubKeyTypeNotSupported,
+				"got: %s, expected: %s", pk.Type(), paramsRes.Params.Validator.PubKeyTypes,
+			)
+		}
+
+		if pkType == sdk.PubKeyEd25519Type && len(pk.Bytes()) != ed25519.PubKeySize {
+			return nil, errorsmod.Wrapf(
+				types.ErrConsensusPubKeyLenInvalid,
+				"got: %d, expected: %d", len(pk.Bytes()), ed25519.PubKeySize,
```

Comment:
we cant assume it will be ed25519, there are chains using secp and will be with bls

Reply:
That's like an extra check we do on ed25519, the others are checked above:

```
		if !slices.Contains(paramsRes.Params.Validator.PubKeyTypes, pkType) {
			return nil, errorsmod.Wrapf(
				types.ErrValidatorPubKeyTypeNotSupported,
				"got: %s, expected: %s", pk.Type(), paramsRes.Params.Validator.PubKeyTypes,
			)
		}
```

Reply:
it makes sense to create an e2e test for this feature in which we send malformed pubkeys and see what happens

Reply:
This is straight copy&paste from CreateValidator, it would be the same behavior as that, if it was ever tested